dali/pipeline/operators/reader/loader/loader.h

```diff
@@ -176,13 +171,18 @@ class Loader {
   virtual void ReadSample(LoadTarget& tensor) = 0;
 
   void PrepareMetadata() {
-    loaded_ = true;
-    PrepareMetadataImpl();
+    std::lock_guard<std::mutex> l(prepare_metadata_mutex_);
```

Comment:
you said we don't care about running this from multiple threads but now we just replaced an std::once_flag by a regular flag and a mutex. Isn't it the same?
I think I prefer the previous approach

Reply:
it turned out that we do - @Kh4L  told me, that you can call size in main thread while training goes (and calls read) in other threads, for example.
In such a case I also think `call_once` is better.

Reply:
That's right

Reply:
`std::call_once` would be interesting if we would be using it without anything else. But here we already rely on loaded_ here, which is lighter.

This `lock_guard` will be reached only one time. Then once `loaded_` is set to true, we won't have  even enter `PrepareMetadata` anymore

Reply:
So I think this bool + mutex condition is what we want to have here

Reply:
But I think we don't need to rely on `loaded_` at all. I added it to produce meaningful error from `Size`. Now `Size` can always be called safely. If call to `PrepareMetadata` is always wrapped with `call_once` we should be good. The only question I have is this: some `PrepareMetadataImpl` implementations were calling `Size`. Now, they would be effectively calling `PrepereMetadataImpl` recursively. How `call_once` acts in such a case? If it works (no recursive call), then we are home and everything should be all right.

Reply:
Now loaded_ is set to true before calling PrepareMetadataImpl, so if you call anything in PrepareMetadataImpl, it will be protected by the bool

Reply:
I don't think I follow. Initially, `loaded_` was set to true after `PrepareMetadataImpl`. Its only intended use was to be checked in `Size` and if false, produce meaningful error. It turned out, that some of the `PrepareMetadataImpl` implementations were calling `Size`. I assumed, that if `Size` is called inside `PrepareMetadataImpl` it is done safely (when it works). I bypassed the problem by moving `loaded_=true` above `PrepareMetadataImple` call.
So my thinking was, that now, we can remove the `loaded_` flag, because we will never be throwing an error from `Size`. Only thing is to make sure that `PrepareMetadataImpl` is not called recursively and is called only once.
Am I wrong here?